python/felis/metadata.py

```diff
@@ -0,0 +1,509 @@
+# This file is part of felis.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+import logging
+from typing import IO, Any, Literal
+
+import sqlalchemy.schema as sqa_schema
+from sqlalchemy import (
+    CheckConstraint,
+    Column,
+    Constraint,
+    Engine,
+    ForeignKeyConstraint,
+    Index,
+    MetaData,
+    Numeric,
+    PrimaryKeyConstraint,
+    ResultProxy,
+    Table,
+    UniqueConstraint,
+    create_mock_engine,
+    make_url,
+    text,
+)
+from sqlalchemy.engine.mock import MockConnection
+from sqlalchemy.engine.url import URL
+from sqlalchemy.exc import SQLAlchemyError
+from sqlalchemy.types import TypeEngine
+
+from felis.datamodel import Schema
+from felis.db._variants import make_variant_dict
+
+from . import datamodel as dm
+from .db import sqltypes
+from .types import FelisType
+
+logger = logging.getLogger(__name__)
+
+
+class InsertDump:
+    """An Insert Dumper for SQL statements which supports writing messages
+    to stdout or a file.
+    """
+
+    def __init__(self, file: IO[str] | None = None) -> None:
```

Comment:
> Does that work? It might be clearer that way in terms of typing because since then `self.file` has no `None` to deal with. If `self.file` is only ever passed to `print()` it might not matter.

This seems to have the unwanted consequence that when `sys.stdout` is provided explicitly, the CLI test methods will print to the terminal rather than suppressing output. I am going to leave it as is for now.

Reply:
Ok. I guess the test code overrides the default output.